nexus/db-queries/src/db/datastore/scim_provider_store.rs

```diff
@@ -7,124 +7,1789 @@
+                    display_name",
+                ),
+            })
+            .collect();
+
+        if tuples.is_empty() {
+            Ok(None)
+        } else {
+            let groups = tuples
+                .into_iter()
+                .map(|column| UserGroup {
+                    // Note neither the scim2-rs crate or Nexus supports nested
+                    // groups
+                    member_type: Some(UserGroupType::Direct),
+                    value: Some(column.group_id.to_string()),
+                    display: Some(column.display_name),
+                })
+                .collect();
+
+            Ok(Some(groups))
+        }
+    }
+
+    async fn get_group_members_for_group_in_txn(
+        &self,
+        conn: &async_bb8_diesel::Connection<DbConnection>,
+        group_id: SiloGroupUuid,
+    ) -> Result<Option<IdOrdMap<GroupMember>>, diesel::result::Error> {
+        use nexus_db_schema::schema::silo_group_membership::dsl;
+
+        let users: Vec<Uuid> = dsl::silo_group_membership
+            .filter(dsl::silo_group_id.eq(to_db_typed_uuid(group_id)))
+            .select(dsl::silo_user_id)
+            .load_async(conn)
+            .await?;
+
+        if users.is_empty() {
+            Ok(None)
+        } else {
+            let members = users
+                .into_iter()
+                .map(|user_id| GroupMember {
+                    // Note neither the scim2-rs crate or Nexus support nested
+                    // groups
+                    resource_type: Some(ResourceType::User.to_string()),
+                    value: Some(user_id.to_string()),
+                })
+                .collect();
+
+            Ok(Some(members))
+        }
+    }
+
+    async fn get_user_by_id_in_txn(
+        &self,
+        conn: &async_bb8_diesel::Connection<DbConnection>,
+        _err: OptionalError<ProviderStoreError>,
+        user_id: SiloUserUuid,
+    ) -> Result<Option<StoredParts<User>>, diesel::result::Error> {
+        let maybe_user = {
+            use nexus_db_schema::schema::silo_user::dsl;
+
+            dsl::silo_user
+                .filter(dsl::silo_id.eq(self.authz_silo.id()))
+                .filter(
+                    dsl::user_provision_type.eq(model::UserProvisionType::Scim),
+                )
+                .filter(dsl::id.eq(to_db_typed_uuid(user_id)))
+                .filter(dsl::time_deleted.is_null())
+                .select(model::SiloUser::as_returning())
+                .first_async(conn)
+                .await
+                .optional()?
+        };
+
+        let Some(user) = maybe_user else {
+            return Ok(None);
+        };
+
+        let groups: Option<Vec<UserGroup>> =
+            self.get_user_groups_for_user_in_txn(conn, user_id).await?;
+
+        let SiloUser::Scim(user) = user.into() else {
+            // With the user provision type filter, this should never be another
+            // type.
+            unreachable!();
```

Comment:
Maybe better to 500 here? This will be a 500, I guess, but doing it explicitly lets you stick a message on it. This could be an argument against using `diesel::result::Error` as your error type for these functions. Most of the datastore functions return `external::Error`.